fix: Issue #30263 with :checkhealth.

sathya-pramodh opened the pull request:
# Description
This PR fixes the issue with `:checkhealth` not recognizing `g:loaded_python3_provider=1` being set in the config (which leads to erroneous behavior as stated in #30263 ).

P.S: This is my first PR in neovim :confetti_ball: so excuse my silliness (if any) :D

zeertzjq commented:
This will report an error when the python3 provider is loaded.

sathya-pramodh commented:
Apparently, my local setup sets `g:loaded_python3_provider=2` ? That's weird.

zeertzjq commented:
Oh wait, `g:loaded_python3_provider` is set to `1` when loading the python3 provider failed. But `:checkhealth` always has warning for the case providers are unavailable.

zeertzjq commented:
IDK, perhaps the providers should instead set their `g:loaded_*_provider` variables to `0` instead of `1` when not available, so that setting the variables to `1` is always wrong?

sathya-pramodh commented:
I wanted to test it out with pyenv like OP said, but silly pip doesn't let me do it XD

sathya-pramodh commented:
I figured it out :D

justinmk commented:
Will revisit this after release. I don't have time to carefully look at this. Please keep it rebased / leave a comment after Nvim 0.11 is released.

sathya-pramodh commented:
Yes! Sure ty! Will do (and fix lint errors)

justinmk commented on runtime/lua/vim/health/health.lua:
```diff
@@ -110,6 +110,16 @@ local function check_config()
     )
   end
 
+  if vim.g.loaded_python3_provider and vim.g.loaded_python3_provider == 1 then
```
comparing `nil` to 1 is valid in Lua

```suggestion
  if vim.g.loaded_python3_provider == 1 then
```

sathya-pramodh replied:
Oh yeah makes sense. Let me push a commit fixing it right now

sathya-pramodh replied:
Fixed in latest commit.

justinmk commented on runtime/lua/vim/health/health.lua:
```diff
@@ -110,6 +110,16 @@ local function check_config()
+  if vim.g.loaded_python3_provider == 1 then
+    ok = false
+    health.error(
+      '`g:loaded_python3_provider=1` may have been set by mistake. This option should not be used to load python provider in your config.',
```
the provider's own code sets 1 here: https://github.com/neovim/neovim/blob/95ab723995f1cc1a11c62355fbac64597526d47e/runtime/autoload/provider/python3.vim#L14

is this message going to be shown in that case?

zeertzjq replied:
Perhaps the provider code should be changed to the set it to 0 instead:

sathya-pramodh replied:
The tests i did were ":echo g:loaded_python3_provider" and ":echo has('python3')" on having the variable set in the config and not. And ":checkhealth" seems to report it properly that it was being set to 2. I wonder if removing the provider still reports it. I will test that right now hold on...

sathya-pramodh replied:
Running ":checkhealth":
![image](https://github.com/user-attachments/assets/83ac34d9-2835-42b3-8b4d-922aed0d68d3)

Running ":echo has('python3")":
![image](https://github.com/user-attachments/assets/2acbf060-fd3e-4818-a3b4-56ffaa036c99)

My init.lua:
![image](https://github.com/user-attachments/assets/6539ba39-098c-45bf-a211-73d0a01ffa8d)

This is on a fresh python venv with no pynvim installed and with the code from the issue branch.

justinmk commented on runtime/lua/vim/health/health.lua:
```diff
@@ -110,6 +110,16 @@ local function check_config()
+    health.error(
+      '`g:loaded_python3_provider=1` may have been set by mistake. This option should not be used to load python provider in your config.',
+      {
+        'Remove the line `vim.g.loaded_python3_provider=1` from your config or comment it out.',
```
```suggestion
        'Remove `vim.g.loaded_python3_provider=1` from your config.',
```

sathya-pramodh replied:
Would these checks still need to be done because we are changing the provider code itself?

justinmk replied:
The purpose of the check is to warn about users accidentally setting `1` in their config

sathya-pramodh replied:
Fixed in latest commit(s)

justinmk replied:
> Would these checks still need to be done because we are changing the provider code itself?

I think the check is still needed because `1` will behave like `0`, which is not what the user intended?

justinmk commented:
> It does detect that loaded_python3_provider being set in the config on my setup

Irrelevant. The provider itself, under certain conditions, sets `1`. The health logic added in this PR, will be misleading under those conditions.

> Perhaps the provider code should be changed to the set it to 0 instead: [#32696 (comment)](https://github.com/neovim/neovim/pull/32696#issuecomment-2692714967)

`1` means the module was "loaded" but the required program wasn't found. However, that information is not actually used anywhere. So technically we could change all places that set 1, to 0:

https://github.com/neovim/neovim/blob/6ef5dd5266822960d254b8acd5ab0934cc2b117e/runtime/autoload/provider/ruby.vim#L14

https://github.com/neovim/neovim/blob/6ef5dd5266822960d254b8acd5ab0934cc2b117e/runtime/autoload/provider/python3.vim#L14

https://github.com/neovim/neovim/blob/6ef5dd5266822960d254b8acd5ab0934cc2b117e/runtime/autoload/provider/perl.vim#L14

https://github.com/neovim/neovim/blob/6ef5dd5266822960d254b8acd5ab0934cc2b117e/runtime/autoload/provider/node.vim#L155

https://github.com/neovim/neovim/blob/6ef5dd5266822960d254b8acd5ab0934cc2b117e/runtime/autoload/provider/clipboard.vim#L287

https://github.com/neovim/neovim/blob/6ef5dd5266822960d254b8acd5ab0934cc2b117e/runtime/autoload/provider/node.vim#L4

https://github.com/neovim/neovim/blob/6ef5dd5266822960d254b8acd5ab0934cc2b117e/runtime/autoload/provider/clipboard.vim#L12

Ok, let's make that change. All of this will become simpler once the providers are migrated to Lua...

sathya-pramodh committed:
feat(checkhealth): ":checkhealth" to report loaded_python3_provider=1

Problem:
":checkhealth" doesn't report `vim.g.loaded_python3_provider = 1`
  under `Configuration` checks.

Full issue discussion here: https://github.com/neovim/neovim/issues/30263.

Solution:
Issue solved by adding a check to see if that variable is being set to `1`
while running ":checkhealth".

sathya-pramodh committed:
fix(provider): providers set loaded_*_provider=0 when empty

justinmk reviewed (approved):
This is probably safe to do. Because any code that was checking for `1` explicitly was already broken, since these values may be set to `2` depending on the current system.

justinmk commented on runtime/lua/vim/health/health.lua:
```diff
@@ -110,6 +110,16 @@ local function check_config()
+  if vim.g.loaded_python3_provider == 1 then
+    ok = false
+    health.error(
+      '`g:loaded_python3_provider=1` may have been set by mistake. This option should not be used to load python provider in your config.',
```
This can be generalized for all the providers. I'll do that in a followup PR.
